crates/sui-adapter/src/adapter.rs

```diff
@@ -480,294 +504,156 @@ fn process_successful_execution<
-        // TODO: Do we charge extra gas when creating new objects (on top of storage write cost)?
+
+    for (id, delete_kind) in deletions {
+        let version = match by_value_objects.get(&id) {
+            Some((_, version)) => *version,
+            None => match state_view.get_latest_parent_entry_ref(id) {
+                Ok(Some((_, previous_version, _))) => previous_version,
+                Ok(None) => {
+                    // TODO we don't really need to mark this as deleted, as the object was not
+                    // created this txn but has never existed in storage. We just need a better
+                    // way of detecting this rather than relying on the parent sync
+                    SequenceNumber::new()
+                }
+                _ => {
+                    return Err(ExecutionError::new_with_source(
+                        ExecutionErrorKind::InvariantViolation,
+                        missing_unwrapped_msg(&id),
+                    ))
+                }
+            },
+        };
+        state_view.log_event(Event::delete_object(
+            module_id.address(),
+            module_id.name(),
+            ctx.sender(),
+            id,
+        ));
+        changes.insert(id, ObjectChange::Delete(version, delete_kind));
     }
-    let obj_address: SuiAddress = obj_id.into();
-    object_owner_map.remove(&obj_address);
-    // increment new owner count, if applicable
-    match recipient {
-        Owner::Shared | Owner::AddressOwner(_) => (),
-        Owner::Immutable => {
-            frozen_object_ids.insert(id);
-        }
-        Owner::ObjectOwner(new_owner) => {
-            // Below we check whether the transfer introduced any circular ownership.
-            // We know that for any mutable object, all its ancenstors (if it was owned by another object)
-            // must be in the input as well. Prior to this we have recorded the original ownership mapping
-            // in object_owner_map. For any new transfer, we trace the new owner through the ownership
-            // chain to see if a cycle is detected.
-            // TODO: Set a constant upper bound to the depth of the new ownership chain.
-            let mut parent = new_owner;
-            while parent != obj_address && object_owner_map.contains_key(&parent) {
-                parent = *object_owner_map.get(&parent).unwrap();
-            }
-            if parent == obj_address {
-                return Err(ExecutionErrorKind::circular_object_ownership(parent.into()).into());
-            }
-            object_owner_map.insert(obj_address, new_owner);
-        }
+
+    for (tag, contents) in user_events {
+        state_view.log_event(Event::move_event(
+            module_id.address(),
+            module_id.name(),
+            ctx.sender(),
+            tag,
+            contents,
+        ))
     }
-    Ok((obj, write_kind))
+
+    // apply object writes and object deletions
+    state_view.apply_object_changes(changes);
```

Comment:
No, the state_view is still transactional. We should probably make the API more clear for that for events.

The reason I did not change it was it didn't do too much. The state_view used to manually manage and infer all of the wraps/creates  etc from imperatively specified deletes/writes. But all of that information was already present in the adapter here, so I changed the API to instead take this single change-set. 

The event log is write only, so the same sort of complexity didn't exist, but we should probably change it as this style is more bug resistant. As I think in that style it will be harder fro the changes to get out of sync with the events.